mava/systems/sable/anakin/rec_sable.py

```diff
@@ -0,0 +1,695 @@
+            learner_state (NamedTuple):
+                - params (FrozenDict): The initial model parameters.
+                - opt_state (OptState): The initial optimizer state.
+                - key (chex.PRNGKey): The random number generator state.
+                - env_state (LogEnvState): The environment state.
+                - timesteps (TimeStep): The initial timestep in the initial trajectory.
+                - hstates (HiddenStates): The initial hidden states of the network.
+
+        """
+        batched_update_step = jax.vmap(_update_step, in_axes=(0, None), axis_name="batch")
+
+        learner_state, (episode_info, loss_info) = jax.lax.scan(
+            batched_update_step, learner_state, None, config.system.num_updates_per_eval
+        )
+        return ExperimentOutput(
+            learner_state=learner_state,
+            episode_metrics=episode_info,
+            train_metrics=loss_info,
+        )
+
+    return learner_fn
+
+
+def learner_setup(
+    env: MarlEnv, keys: chex.Array, config: DictConfig
+) -> Tuple[LearnerFn[LearnerState], Callable, LearnerState]:
+    """Initialise learner_fn, network, optimiser, environment and states."""
+    # Get available TPU cores.
+    n_devices = len(jax.devices())
+
+    # Get number of agents.
+    config.system.num_agents = env.num_agents
+
+    # PRNG keys.
+    key, net_key = keys
+
+    # Get number of agents and actions.
+    action_dim = int(env.action_spec().num_values[0])
+    n_agents = env.action_spec().shape[0]
+    config.system.num_agents = n_agents
+    config.system.num_actions = action_dim
+
+    _, action_space_type = get_action_head(env)
+
+    # Define network.
+    sable_network = SableNetwork(
+        n_agents=n_agents,
+        action_dim=action_dim,
+        rollout_length=config.system.rollout_length,
+        net_config=config.network.net_config,
+        memory_config=config.network.memory_config,
+        action_space_type=action_space_type,
+    )
+
+    # Define optimiser.
+    lr = make_learning_rate(config.system.actor_lr, config)
+    optim = optax.chain(
+        optax.clip_by_global_norm(config.system.max_grad_norm),
+        optax.adam(lr, eps=1e-5),
+    )
+
+    # Get mock inputs to initialise network.
+    init_obs = env.observation_spec().generate_value()
+    init_obs = tree.map(lambda x: x[jnp.newaxis, ...], init_obs)  # Add batch dim
+    init_hs = get_init_hidden_state(config.network.net_config, config.arch.num_envs)
+    init_hs = tree.map(lambda x: x[0, jnp.newaxis], init_hs)
+
+    # Initialise params and optimiser state.
+    params = sable_network.init(
+        net_key,
+        init_obs,
+        init_hs,
+        net_key,
+        method="init_net",
+    )
+    opt_state = optim.init(params)
+
+    # Pack apply and update functions.
+    apply_fns = (
+        partial(sable_network.apply, method="get_actions"),  # Execution function
+        sable_network.apply,  # Training function
+    )
+    update_fn = optim.update
```

Comment:
We can just use this directly now instead of having to define an extra variable here.